### PR TITLE
[SPARK-48056][CONNECT][PYTHON] Re-execute plan if a SESSION_NOT_FOUND error is raised and no partial response was received

### PR DESCRIPTION
**Pull request by nija-at**
### What changes were proposed in this pull request?

Similar to OPERATION_NOT_FOUND, re-attempt to execute
the original spark connect plan when a SESSION_NOT_FOUND is
received from the spark connect service and no partial responses
were previously received.

### Why are the changes needed?

This error has been noticed to occur during a cluster cold start
and when a request arrives when the connect service is not fully
initialized.

### Does this PR introduce _any_ user-facing change?

Prevoiusly, connect-based pyspark APIs would fail with the error code
"INVALID_HANDLE.SESSION_NOT_FOUND" in the very first request to
the service.
With this change, the client will now automatically retry.

### How was this patch tested?

Attached unit tests.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: python/pyspark/sql/tests/connect/client/test_client.py [hunk 1]
```diff
@@ -340,6 +353,71 @@ def check():
...
+        eventually(timeout=1, catch_assertions=True)(checks)()
+
+    @parameterized.expand(
+        [
+            ("session", "INVALID_HANDLE.SESSION_NOT_FOUND"),
+            ("operation", "INVALID_HANDLE.OPERATION_NOT_FOUND"),
+        ]
+    )
+    def test_not_found_fails(self, _, error_msg: str):
+        # Assert that the client fails from session or operation not found error
```
Comment (grundprinzip) [inline python/pyspark/sql/tests/connect/client/test_client.py:394]: nit: use """ """ for documentation of the test ideally something along the lines of this:

```suggestion
        """ SPARK-48056: Assert that the client fails from session or operation not found error
        """
```

File: python/pyspark/sql/tests/connect/client/test_client.py [hunk 2]
```diff
@@ -340,6 +353,71 @@ def check():
...
         eventually(timeout=1, catch_assertions=True)(check)()
 
+    @parameterized.expand(
+        [
+            ("session", "INVALID_HANDLE.SESSION_NOT_FOUND"),
+            ("operation", "INVALID_HANDLE.OPERATION_NOT_FOUND"),
+        ]
+    )
+    def test_not_found_recovers(self, _, error_msg: str):
+        # Assert that the client recovers from session or operation not found error
```
Comment (grundprinzip) [inline python/pyspark/sql/tests/connect/client/test_client.py:363]: ame nit

**Review by grundprinzip — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Review by juliuszsompolski — COMMENTED** [review]
Could you fix the same in scala client in https://github.com/apache/spark/blob/master/connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/ExecutePlanResponseReattachableIterator.scala#L237 ?

**Comment by nija-at** [comment]
@juliuszsompolski - this will be followed up in a separate PR

**Review by juliuszsompolski — APPROVED** [review]

File: dev/requirements.txt [hunk 3]
```diff
@@ -16,6 +16,7 @@ memory-profiler>=0.61.0
 # PySpark test dependencies
 unittest-xml-reporting
 openpyxl
+parameterized
```
Comment (grundprinzip) [inline dev/requirements.txt:19]: please pin the version

File: dev/infra/Dockerfile [hunk 4]
```diff
@@ -89,7 +89,7 @@ RUN curl -sS https://bootstrap.pypa.io/get-pip.py | pypy3
 RUN pypy3 -m pip install numpy 'six==1.16.0' 'pandas==2.2.2' scipy coverage matplotlib lxml
 
 
-ARG BASIC_PIP_PKGS="numpy pyarrow>=15.0.0 six==1.16.0 pandas==2.2.2 scipy plotly>=4.8 mlflow>=2.8.1 coverage matplotlib openpyxl memory-profiler>=0.61.0 scikit-learn>=1.3.2"
+ARG BASIC_PIP_PKGS="numpy pyarrow>=15.0.0 six==1.16.0 pandas==2.2.2 scipy plotly>=4.8 mlflow>=2.8.1 coverage matplotlib openpyxl memory-profiler>=0.61.0 scikit-learn>=1.3.2 parameterized"
```
Comment (grundprinzip) [inline dev/infra/Dockerfile:92]: same here and in the workflow config

**Review by grundprinzip — COMMENTED** [review]
I don't have a problem with adding the dependency, but we should make sure that it's only needed for testing and documented in the requirements file as such.

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/sql/tests/connect/client/test_client.py [hunk 5]
```diff
@@ -18,13 +18,15 @@
 import unittest
 import uuid
 from collections.abc import Generator
-from typing import Optional, Any
+from typing import Optional, Any, Union
+from parameterized import parameterized
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/connect/client/test_client.py:22]:
```suggestion
```

File: dev/requirements.txt (hunk 3 above)
Comment (nija-at) [reply dev/requirements.txt]: removed the new dep

File: dev/infra/Dockerfile (hunk 4 above)
Comment (nija-at) [reply dev/infra/Dockerfile]: removed the new dep

**Comment by HyukjinKwon** [comment]
Merged to master.

**Review by dongjoon-hyun — COMMENTED** [review]
Hi, @nija-at , @grundprinzip, @juliuszsompolski , @HyukjinKwon .

I made a follow-up to fix Python 3.12 issue. `assertEquals` is removed at Python 3.12.
- https://github.com/apache/spark/pull/46377

**Comment by dongjoon-hyun** [comment]
This seems to be missed in another PR too.
- #45242

**Comment by HyukjinKwon** [comment]
Thanks for fixing

**Comment by changgyoopark-db** [comment]
> Could you fix the same in scala client in https://github.com/apache/spark/blob/master/connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/ExecutePlanResponseReattachableIterator.scala#L237 ?

I just uploaded a scala version of it: https://github.com/apache/spark/pull/46971, could you review this?

**Pull request closed without merging**